sys/include/net/gnrc/tcp.h

```diff
@@ -34,13 +34,62 @@
 extern "C" {
 #endif
 
+
+/**
+ * @brief Address information for a single TCP connection endpoint.
+ * @extends sock_tcp_ep_t
+ */
+typedef struct {
+    int family;                            /**< IP Address family. */
+#ifdef MODULE_GNRC_IPV6
+    union {
+        uint8_t ipv6[sizeof(ipv6_addr_t)]; /**< IPv6 Address storage */
+    } addr;                                /**< IP Address storage */
+#endif
```

Comment:
If you want to make it transparent to `sock_tcp_ep_t` (as you documented above), why not do it as we do there:

```C
    union {
#if IS_USED(MODULE_GNRC_IPV6)
        /**
         * @brief IPv6 address mode
         *
         * @note only available with module `gnrc_ipv6`.
         */
        uint8_t ipv6[sizeof(ipv6_addr_t)];
#endif
        /**
         * @brief IPv4 address storage
         */
        uint8_t ipv4[sizeof(ipv4_addr_t];
        uint32_t ipv4_u32;      /**< IPv4 address *in network byte order* */
    } addr;                 /**< address */
```

Yes, we do not support IPv4 (yet... see #12988), but your check for the `family` field should make sure people do not use the IPv4 address "accidentally".

Reasoning for using IPv4 address as a placeholder: it is at least a word long but shorter than an IPv6 address (otherwise you could just keep the IPv6 address in). Anything shorter will most likely lead to padding anyway.

Reply:
I did not add IPv4 storage because currently it gnrc does not support it.
If I use a exact reimplementation of sock_ep_t, I would have to use ipv4_addr_t (not available in gnrc). Since I want to keep gnrc_tcp independent from sock I could use a magic number instead of ipv4_addr_t but I would like to keep it as it is.

The current implementation is compatible with the sock struct if IPv6 is used, otherwise a operation would not work anyway.

Reply:
Could you clarify:

> Yes, we do not support IPv4 (yet... see #12988), but your check for the family field should make sure people do not use the IPv4 address "accidentally".

I don't really understand what you mean.

Reply:
Oh wait this wasn't resolved...

You have e.g. [this check](https://github.com/RIOT-OS/RIOT/pull/13026/files#diff-37f9e18ba67907d8e3d0a61e4f8416f4R148-R157). If that is not given you return `-EAFNOSUPPORT` (including when `!IS_USED(MODULE_GNRC_IPV6)`), right?

Reply:
Your link points me only to the interface. The implementation of "gnrc_tcp_ep_init" has this check:

```
#ifdef MODULE_GNRC_IPV6
    if (family != AF_INET6) {
        return -EAFNOSUPPORT;
    }

    if (addr == NULL && addr_size == 0) {
        ipv6_addr_set_unspecified((ipv6_addr_t *) ep->addr.ipv6);
    }
    else if (addr_size == sizeof(ipv6_addr_t)) {
        memcpy(ep->addr.ipv6, addr, sizeof(ipv6_addr_t));
    }
    else {
        return -EINVAL;
    }
#else
    return -EAFNOSUPPORT;
#endif
```

So in any case it should return -EAFNOSUPPORT if anything else than AF_INET6 is supplied.

Reply:
Mh, shouldn't the PCB create-functions also check for that? Nobody prevents a user from just filling the `ep` themselves.

Reply:
Let's see it the other way around: is there any way that TCP is used without IP? I don't believe so. So you need at least some address and the smallest common denominator is an IPv4 address.

Reply:
The connection open functions check for the address family. Event if the endpoint is constructed by hand, it is checked on connection opening.

Originally gnrc_tcp was able to be compilable without a specified network layer (it asserted on operation), I tested if this still works, it is no longer possible.

Could address this on two ways:

1) Ensure that all operations are within a #ifdef MODULE_GNRC_IPV6 block
2) Add Dummy IPv4 Address (although ipv4_addr_t is not available in gnrc, so it might be an array with a magic number)

I prefer to do it the first way, do you have any objections before I start fixing this?

Reply:
Regarding 2. I don't really understand where the magic number is coming from in your opinion. All I want is a field the `union` can fall back on (let's call it a dummy if you insist) if IPv6 is not available. As your init-function is checking the family, it is not initialized via that and as the open functions check the family as well, the value is not accessed there either. So no numbers involved. When the `union` is already there we have to change `gnrc_sock_tcp` only minimally (or even better just add to it) once IPv4 is there. The `#ifdef MODULE_GNRC_IPV6` block you will need either way.

Reply:
Ah now I get what you mean. I'll add the changes.

Reply:
I wrote down in the [first comment](https://github.com/RIOT-OS/RIOT/pull/13026#discussion_r382677467) what I wanted ;-)

Reply:
Remote collaboration is sometimes lossy ;)

The latest commit should contain resolve the Issue. If you are happy with it I'll squash.